Failed test look a bit suspicious. Complaining about read-only mount. I've added one, but it is only under `--debug`
```
E               Exception: Command [docker exec rootteststoragedelta-gw0-node1-1 bash -c mkdir -p $(dirname /ClickHouse/tests/integration/test_storage_delta/_instances-0-gw0/node1/database/user_files/test_checkpointlHzqpHKQqs/part-00000-6b703aa8-2103-4f68-b475-7d97e6b42f5c-c000.parquet) && echo UEFSMRUAFRwVHBXEu8jVCxwVAhUAFQYVCAAAAgAAAAMBCgAAAAAAAAAVABUYFRgVyMjv1QwcFQIVABUGFQgAAAIAAAADAQIAAAAxMRkRAhkYCAoAAAAAAAAAGRgICgAAAAAAAAAVAhkWAAAZEQIZGAIxMRkYAjExFQIZFgAAGRwWCBVKFgAAABkcFlIVRhYAAAAVAhk8SAxzcGFya19zY2hlbWEVBAAVBCUCGChjb2wtNDBiN2Q0NmMtZTRmZS00Mjc4LWFmOGMtZjg2OTI5ZWUxMWEyVQIAFQwlAhgoY29sLTlkNzY5NmIyLTMxM2EtNDViYy1iMmNiLTNiN2UzYzEyZjBlNCUANQQcHAAAABYCGRwZLCYIHBUEGTUIAAYZGChjb2wtNDBiN2Q0NmMtZTRmZS00Mjc4LWFmOGMtZjg2OTI5ZWUxMWEyFQAWAhZKFkomCDwYCAoAAAAAAAAAGAgKAAAAAAAAABYAKAgKAAAAAAAAABgICgAAAAAAAAAAGRwVABUAFQIAABb8ARUUFpgBFT4AJlIcFQwZNQgABhkYKGNvbC05ZDc2OTZiMi0zMTNhLTQ1YmMtYjJjYi0zYjdlM2MxMmYwZTQVABYCFkYWRiZSPBgCMTEYAjExFgAoAjExGAIxMQAZHBUAFQAVAgAAFpACFRQW1gEVJgAWkAEWAiYIFpABFAAAGSwYGG9yZy5hcGFjaGUuc3BhcmsudmVyc2lvbhgFMy41LjUAGClvcmcuYXBhY2hlLnNwYXJrLnNxbC5wYXJxdWV0LnJvdy5tZXRhZGF0YRjYA3sidHlwZSI6InN0cnVjdCIsImZpZWxkcyI6W3sibmFtZSI6ImNvbC00MGI3ZDQ2Yy1lNGZlLTQyNzgtYWY4Yy1mODY5MjllZTExYTIiLCJ0eXBlIjoibG9uZyIsIm51bGxhYmxlIjp0cnVlLCJtZXRhZGF0YSI6eyJkZWx0YS5jb2x1bW5NYXBwaW5nLmlkIjoxLCJkZWx0YS5jb2x1bW5NYXBwaW5nLnBoeXNpY2FsTmFtZSI6ImNvbC00MGI3ZDQ2Yy1lNGZlLTQyNzgtYWY4Yy1mODY5MjllZTExYTIiLCJwYXJxdWV0LmZpZWxkLmlkIjoxfX0seyJuYW1lIjoiY29sLTlkNzY5NmIyLTMxM2EtNDViYy1iMmNiLTNiN2UzYzEyZjBlNCIsInR5cGUiOiJzdHJpbmciLCJudWxsYWJsZSI6dHJ1ZSwibWV0YWRhdGEiOnsiZGVsdGEuY29sdW1uTWFwcGluZy5pZCI6MiwiZGVsdGEuY29sdW1uTWFwcGluZy5waHlzaWNhbE5hbWUiOiJjb2wtOWQ3Njk2YjItMzEzYS00NWJjLWIyY2ItM2I3ZTNjMTJmMGU0IiwicGFycXVldC5maWVsZC5pZCI6Mn19XX0AGEpwYXJxdWV0LW1yIHZlcnNpb24gMS4xMy4xIChidWlsZCBkYjQxODMxMDlkNWI3MzRlYzU5MzBkODcwY2RhZTE2MWU0MDhkZGJhKRksHAAAHAAAAPgDAABQQVIx | base64 --decode > /ClickHouse/tests/integration/test_storage_delta/_instances-0-gw0/node1/database/user_files/test_checkpointlHzqpHKQqs/part-00000-6b703aa8-2103-4f68-b475-7d97e6b42f5c-c000.parquet] return non-zero code 1: bash: line 1: /ClickHouse/tests/integration/test_storage_delta/_instances-0-gw0/node1/database/user_files/test_checkpointlHzqpHKQqs/part-00000-6b703aa8-2103-4f68-b475-7d97e6b42f5c-c000.parquet: Read-only file system
```

--- a/tests/integration/README.md
+++ b/tests/integration/README.md
@@ -227,7 +227,7 @@ will automagically detect the types of variables and only the small diff of two 
 
 Here is how you run a debugger for the server under integration test:
 
-1. Put the **_statically linked binary_** of your debugger into the ClickHouse root folder of your repo. The [nnd](https://github.com/al13n321/nnd) debugger is ideal for this purpose (and for ClickHouse debugging in general).
+1. Put the **_statically linked binary_** of your debugger into the ClickHouse root folder of your repo. The [nnd](https://github.com/al13n321/nnd) debugger is ideal for this purpose (and for ClickHouse debugging in general), it's currently the only supported option.
 2. Go to the integration test `test.py` file and add a new line with a single `breakpoint()` command to debug specific point in your test (to make the server work forever and not die after the test).
 3. Run the integration test with `runner` script as usual but add `--debug` option. It will produce helper shell command in stdout and start testing:
 ```bash

--- a/tests/integration/README.md
+++ b/tests/integration/README.md
@@ -229,20 +229,13 @@ Here is how you run a debugger for the server under integration test:
 
 1. Put the **_statically linked binary_** of your debugger into the ClickHouse root folder of your repo. The [nnd](https://github.com/al13n321/nnd) debugger is ideal for this purpose (and for ClickHouse debugging in general).
 2. Go to the integration test `test.py` file and add a new line with a single `breakpoint()` command to debug specific point in your test (to make the server work forever and not die after the test).
-3. Run the integration test with `runner` script as usual but add `--debug` option. It will produce helper shell commands in stdout and start testing:
+3. Run the integration test with `runner` script as usual but add `--debug` option. It will produce helper shell command in stdout and start testing:
 ```bash
-    # =====> DEBUG MODE <=====
-    # ClickHouse root folder will be read=only mounted into /ClickHouse in all containers.
-    # Use the following command to list node containers.
-    runner-nodes() {
-        docker exec clickhouse_integration_tests_160lfw docker ps | grep node
-    }
-
-    # You can use the following command to log into the container and run clickhouse client or debugger.
-    # Tip: place a `breakpoint()` somewhere in your integration test python code before run.
-    runner-bash() {
-        docker exec -it clickhouse_integration_tests_160lfw docker exec -it $1 bash
-    }
+   # =====> DEBUG MODE <=====
+    # ClickHouse root folder will be read-only mounted into /debug in all containers.
+    # Tip: place a `breakpoint()` somewhere in your integration test python code before `runner`.
+    # Open another shell and include:
+        source /path/to/ClickHouse/tests/integration/runner-env.sh
     # =====> DEBUG MODE <=====
 ```
 4. When the test hits a breakpoint, it will show the Python debugger prompt, which is useful by itself for debugging purposes. For example:
@@ -253,18 +246,49 @@ test_throttling/test.py::test_write_throttling[user_remote_throttling]
 -> assert_took(took, should_take)
 (Pdb)
 ```
-5. Open another shell and paste the DEBUG MODE commands there to set up the environment. Then run:
+5. Open another shell and run the command there to set up the environment. Then run:
 ```bash
-❯ runner-nodes
-2e2a25260eae   clickhouse/integration-test:latest         "bash -c 'trap 'pkil…"   4 minutes ago   Up 4 minutes                               roottestthrottling-node-1
+❯ source /path/to/ClickHouse/tests/integration/runner-env.sh
+
+USAGE:
+   runner-client - Run clickhouse client inside an integration test
+   runner-bash   - Open shell on a node inside an integration test
+   runner-nnd    - Attach nnd debugger to a clickhouse server on a node inside an integration test
 ```
 6. You can use either Container ID or Container Name to log into the required node:
 ```
-❯ runner-bash roottestthrottling-node-1
-root@node:/#
+❯ runner-client
+CONTAINER ID   IMAGE                                      COMMAND                  CREATED          STATUS          PORTS                                                            NAMES
+867c67cc9957   clickhouse/integration-test:latest         "bash -c 'trap 'pkil…"   2 seconds ago    Up 1 second                                                                      rootteststoragedelta-node1-1
+64b8744cb71a   clickhouse/integration-test:latest         "bash -c 'trap 'pkil…"   2 seconds ago    Up 1 second                                                                      rootteststoragedelta-node2-1
+414187056a06   clickhouse/clickhouse-server:25.3.3.42     "bash -c 'trap 'pkil…"   2 seconds ago    Up 1 second     8123/tcp, 9000/tcp, 9009/tcp                                     rootteststoragedelta-node_old-1
+f5c9aef75a04   clickhouse/integration-test:latest         "clickhouse server -…"   2 seconds ago    Up 1 second                                                                      rootteststoragedelta-node_with_environment_credentials-1
+c7eae7f0c29d   mcr.microsoft.com/azure-storage/azurite    "docker-entrypoint.s…"   21 seconds ago   Up 20 seconds   10000-10002/tcp, 0.0.0.0:30000->30000/tcp, :::30000->30000/tcp   rootteststoragedelta-azurite1-1
+8a81755f9358   minio/minio:RELEASE.2024-07-31T05-46-26Z   "/usr/bin/docker-ent…"   22 seconds ago   Up 21 seconds   9000-9001/tcp                                                    rootteststoragedelta-minio1-1
+6f11a018f16c   clickhouse/python-bottle:latest            "python3"                22 seconds ago   Up 21 seconds   8080/tcp                                                         rootteststoragedelta-resolver-1
+bf180f6b2f7d   clickhouse/s3-proxy                        "/docker-entrypoint.…"   22 seconds ago   Up 22 seconds   80/tcp, 443/tcp, 8080/tcp                                        rootteststoragedelta-proxy1-1
+0c03e8d780e5   clickhouse/s3-proxy                        "/docker-entrypoint.…"   22 seconds ago   Up 22 seconds   80/tcp, 443/tcp, 8080/tcp                                        rootteststoragedelta-proxy2-1
+1ef790b3251e   clickhouse/integration-test:latest         "clickhouse keeper -…"   28 seconds ago   Up 27 seconds                                                                    rootteststoragedelta-zoo2-1
+a65b504a1e60   clickhouse/integration-test:latest         "clickhouse keeper -…"   28 seconds ago   Up 27 seconds                                                                    rootteststoragedelta-zoo3-1
+51d20a4f40c4   clickhouse/integration-test:latest         "clickhouse keeper -…"   28 seconds ago   Up 28 seconds                                                                    rootteststoragedelta-zoo1-1
+Enter ClickHouse Node CONTAINER ID or NAME (default: 867c67cc9957):
+ClickHouse client version 25.7.1.1.
+Connecting to localhost:9000 as user default.
+Connected to ClickHouse server version 25.7.1.
+node1 :) select 1
+
+SELECT 1
+
+Query id: 7222639b-6d33-4237-85f4-ad4ee04e8691
+
+   ┌─1─┐
+1. │ 1 │
+   └───┘
+
+1 row in set. Elapsed: 0.001 sec.
+
+node1 :) Bye.
 ```
-7. Run `clickhouse client` to connect to the server running in the node container.
-8. Find the PID with `ps aux` and run the debugger in the the node container with `/ClickHouse/nnd -p 767 -d /ClickHouse`
 
 ### Troubleshooting
 

--- a/tests/integration/helpers/cluster.py
+++ b/tests/integration/helpers/cluster.py
@@ -3634,7 +3634,7 @@ services:
             - {db_dir}:/var/lib/clickhouse/
             - {logs_dir}:/var/log/clickhouse-server/
             - /etc/passwd:/etc/passwd:ro
-            - /ClickHouse:/ClickHouse:ro
+            - /debug:/debug:ro
             {binary_volume}
             {external_dirs_volumes}
             {odbc_ini_path}

--- a/tests/integration/runner
+++ b/tests/integration/runner
@@ -362,7 +362,7 @@ if __name__ == "__main__":
         action="store_true",
         dest="debug",
         default=False,
-        help="Mount ClickHouse source folder from --clickhouse-root into /ClickHouse (read-only)",
+        help="Mount ClickHouse source folder from --clickhouse-root into /debug (read-only)",
     )
 
     parser.add_argument(
@@ -475,7 +475,7 @@ if __name__ == "__main__":
     )
 
     # Mount debug root into container if debug flag is set
-    root_mount = f"--volume={CLICKHOUSE_ROOT}:/ClickHouse:ro" if args.debug else ""
+    debug_mount = f"--volume={CLICKHOUSE_ROOT}:/debug:ro" if args.debug else ""
 
     cmd_base = (
         f"docker run {net} {tty} --rm --name {CONTAINER_NAME} "
@@ -485,7 +485,7 @@ if __name__ == "__main__":
         f"--cap-add=SYS_PTRACE "
         f"--volume={args.binary}:/clickhouse "
         f"--volume={args.base_configs_dir}:/clickhouse-config "
-        f"{root_mount} "
+        f"{debug_mount} "
         f"--volume={args.cases_dir}:/ClickHouse/tests/integration "
         f"--volume={args.utils_dir}/backupview:/ClickHouse/utils/backupview "
         f"--volume={args.utils_dir}/grpc-client/pb2:/ClickHouse/utils/grpc-client/pb2 "
@@ -519,17 +519,10 @@ if __name__ == "__main__":
         print(
             f"""
     # =====> DEBUG MODE <=====
-    # ClickHouse root folder will be read=only mounted into /ClickHouse in all containers.
-    # Use the following command to list node containers.
-    runner-nodes() {{
-        docker exec {CONTAINER_NAME} docker ps | grep node
-    }}
-
-    # You can use the following command to log into the container and run clickhouse client or debugger.
-    # Tip: place a `breakpoint()` somewhere in your integration test python code before run.
-    runner-bash() {{
-        docker exec -it {CONTAINER_NAME} docker exec -it $1 bash
-    }}
+    # ClickHouse root folder will be read-only mounted into /debug in all containers.
+    # Tip: place a `breakpoint()` somewhere in your integration test python code before `runner`.
+    # Open another shell and include:
+        source {CLICKHOUSE_ROOT}/tests/integration/runner-env.sh
     # =====> DEBUG MODE <=====
 """
         )

--- a/tests/integration/runner-env.sh
+++ b/tests/integration/runner-env.sh
@@ -1,0 +1,41 @@
+#!/bin/zsh
+# Works with bash and zsh
+
+# It should be used with `runner --debug`
+# Tip: place a `breakpoint()` somewhere in your integration test python code before run.
+
+_runner-select-node() {
+    RUNNER_ID1=$(docker ps | grep clickhouse_integration_tests | awk '{print $1; exit}')
+    RUNNER_ID2_DEFAULT=$(docker exec $RUNNER_ID1 docker ps | grep node | awk '{print $1; exit}')
+    docker exec $RUNNER_ID1 docker ps
+    RUNNER_ID2=
+    echo -n "Enter ClickHouse Node CONTAINER ID or NAME (default: $RUNNER_ID2_DEFAULT): "
+    read RUNNER_ID2
+    RUNNER_ID2=${RUNNER_ID2:-$RUNNER_ID2_DEFAULT}
+}
+
+# Runs a clickhouse client for a node inside an integration test.
+runner-client() {
+    _runner-select-node
+    docker exec -it $RUNNER_ID1 docker exec -it $RUNNER_ID2 clickhouse client
+}
+
+# Opens shell on a node inside an integration test.
+runner-bash() {
+    _runner-select-node
+    docker exec -it $RUNNER_ID1 docker exec -it $RUNNER_ID2 bash
+}
+
+# Attaches nnd debugger to a clickhouse server on a node inside an integration test.
+# You have to put statically-linked nnd binary into ClickHouse root folder
+runner-nnd() {
+    _runner-select-node
+    RUNNER_PID=$(docker exec $RUNNER_ID1 docker exec $RUNNER_ID2 sh -c 'ps aux | grep -v grep | grep -v bash | grep -m 1 "clickhouse server" | awk "{print \$2}"')
+    docker exec -it $RUNNER_ID1 docker exec -it $RUNNER_ID2 /debug/nnd -p $RUNNER_PID -d /debug
+}
+
+echo "USAGE:"
+echo "   runner-client - Run clickhouse client inside an integration test"
+echo "   runner-bash   - Open shell on a node inside an integration test"
+echo "   runner-nnd    - Attach nnd debugger to a clickhouse server on a node inside an integration test"
+echo ""

